File: Maryland-Cryptography/Programming_Assignment_3/proj3/oracle.c
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <stdlib.h>
#include <unistd.h>
#include <stdio.h>
#include <string.h>

#define NOFLAGS 0
#define BLOCK_LENGTH 16

extern int sockfd;

int Oracle_Connect() {
  struct sockaddr_in servaddr, cliaddr;
  int rc;

  sockfd = socket(AF_INET, SOCK_STREAM, 0);

  bzero(&servaddr, sizeof(servaddr));
  servaddr.sin_family = AF_INET;
  servaddr.sin_addr.s_addr=inet_addr("54.165.60.84");
  servaddr.sin_port=htons(80);

  if(!connect(sockfd, (struct sockaddr *)&servaddr, sizeof(servaddr))) {
    printf("Connected to server successfully.\n");
    return 0;
  } else {
    perror("Failed to connect to oracle");
    return -1;
  }
}

int Oracle_Disconnect() {
  if(!close(sockfd)) {
    printf("Connection closed successfully.\n");
    return 0;
  } else {
    perror("[WARNING]: You haven't connected to the server yet");
    return -1;
  }
}

// Packet Structure: < num_blocks(1) || ciphertext(16*num_blocks) || null-terminator(1) >
int Oracle_Send(unsigned char* ctext, int num_blocks) {
  int ctext_len = num_blocks * BLOCK_LENGTH;
  unsigned char message[(ctext_len)+2];
  char recvbit[2];

  message[0] = num_blocks;
  memcpy((message+1), ctext, ctext_len);
  message[ctext_len+1] = '\0';

  if(!send(sockfd, message, ctext_len+2, NOFLAGS)) {
    perror("[WARNING]: You haven't connected to the server yet");
    return -1;
  }
  if(!recv(sockfd, recvbit, 2, NOFLAGS)) {
    perror("[ERROR]: Recv failed");
    return -1;
  }
  recvbit[1] = '\0';

  return atoi(recvbit);
}

File: Maryland-Cryptography/Programming_Assignment_3/proj3/oracle.py
s = None

def Oracle_Connect():
    import socket
    global s
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    try:
        s.connect(('54.165.60.84', 80))
    except socket.error as e:
        print e
        return -1

    print "Connected to server successfully."

    return 0

def Oracle_Disconnect():
    if not s:
        print "[WARNING]: You haven't connected to the server yet."
        return -1

    s.close()
    print "Connection closed successfully."

    return 0

# Packet Structure: < num_blocks(1) || ciphertext(16*num_blocks) || null-terminator(1) >
def Oracle_Send(ctext, num_blocks):
    if not s:
        print "[WARNING]: You haven't connected to the server yet."
        return -1

    msg = ctext[:]
    msg.insert(0, num_blocks)
    msg.append(0)

    s.send(bytearray(msg))
    recvbit = s.recv(2)

    try:
        return int(recvbit)
    except ValueError as e:
        return int(recvbit[0])
